frame/contracts/COMPLEXITY.md

```diff
@@ -187,6 +188,20 @@ implementation they just involve a DB read.
 For subsequent calls and instantiations during contract execution, the initialization requires no
 expensive operations.
 
+## Terminate
+
+This function performs the following steps:
+
+1. Check the calling contract is not already on the callstack by calling `is_live`.
+2. `transfer` funds from caller to the beneficiary.
+3. Flag the caller contract as deleted in the overlay.
+
+`is_live` does not do any database access nor does it allocate memory. It walks up the call
+stack and therefore executes in linear time depending on size of the call stack.
+
+**complexity**: Database accesses as described in Transfer + 1 DB write that removes
```

Comment:
The original text stated that it required a number of writes proportional to the number of the key value pairs that the contract had at the moment of removal.

That is not sound and the thinking has changed since then. So we are are using now child tries for storing the storage of a contract. To remove it we kill it. For now, this operation requires iteration over each value. However, in future we expect this to be constant-time (from the perspective of the runtime).

I think we can just explicitly state here that we are using child trie removal and we expect that it will become constant time (from PoV of the runtime)

Reply:
Uff. Corrected the statement. Which upcoming changes will make this constant time?

Reply:
I think we don't have an issue for that, but it seems we've got a PR now https://github.com/paritytech/substrate/pull/5280